graph/hopcroft_karp.cpp

```diff
@@ -0,0 +1,269 @@
+ * 
+ * ###Bipartite graph :
+ * A bipartite graph (or bigraph) is a graph whose vertices can be divided into two disjoint 
+ * and independent sets U and V such that every edge connects a vertex in U to one in V. 
+ * Vertex sets U and V are usually called the parts of the graph. 
+ * Equivalently, a bipartite graph is a graph that does not contain any odd-length cycles.
+ * 
+ * ###Matching and Not-Matching edges :
+ * Given a matching M, edges that are part of matching are called Matching edges and edges that are not part 
+ * of M (or connect free nodes) are called Not-Matching edges.
+ * 
+ * ###Maximum cardinality matching :
+ * Given a bipartite graphs G = ( V = ( X , Y ) , E ) whose partition has the parts X and Y, 
+ * with E denoting the edges of the graph, the goal is to find a matching with as many edges as possible. 
+ * Equivalently, a matching that covers as many vertices as possible.
+ * 
+ * ###Augmenting paths :
+ * Given a matching M, an augmenting path is an alternating path that starts from and ends on free vertices. 
+ * All single edge paths that start and end with free vertices are augmenting paths.
+ * 
+ * 
+ * ###Concept :
+ * A matching M is not maximum if there exists an augmenting path. It is also true other way,
+ * i.e, a matching is maximum if no augmenting path exists.
+ * 
+ * 
+ * ###Algorithm :
+ * 1) Initialize the Maximal Matching M as empty.
+ * 2) While there exists an Augmenting Path P
+ *   Remove matching edges of P from M and add not-matching edges of P to M
+ *   (This increases size of M by 1 as P starts and ends with a free vertex
+ *   i.e. a node that is not part of matching.)
+ * 3) Return M. 
+ * 
+ * 
+ *
+ * @author [Krishna Pal Deora](https://github.com/Krishnapal4050)
+ * 
+ */
+
+
+#include <iostream>
+#include <cstdlib> 
+#include <queue>
+#include <list>
+#include <climits>
+#include <memory>
+
+/**
+ * @brief Represents Bipartite graph for
+ * Hopcroft Karp implementation
+ */
+class BGraph
+{
+	
+    int m;  ///< m is the number of vertices on left side of Bipartite Graph
+    int n;  ///< n is the number of vertices on right side of Bipartite Graph
+    const int NIL;
+    const int INF;
+
+	
+    std::vector<std::list<int> >adj;  ///< adj[u] stores adjacents of left side and 0 is used for dummy vertex
+
+    std::vector<int> pair_u; ///< value of vertex 'u' ranges from 1 to m
+    std::vector<int> pair_v; ///< value of vertex 'v' ranges from 1 to n
+    std::vector<int> dist;   ///< dist represents the distance between vertex 'u' and vertex 'v'
+
+public:
+    BGraph();		       ///<Default Constructor
+    BGraph(int m, int n);     ///< Constructor
+    void addEdge(int u, int v); ///< To add edge
+    
+    bool bfs(); ///< Returns true if there is an augmenting path    
+    bool dfs(int u); ///< Adds augmenting path if there is one beginning with u  
+	
+    int hopcroftKarpAlgorithm();  ///< Returns size of maximum matching
+};
+
+
+/**
+ * Function documentation
```

Comment:
What does this function do? Please document it here.
Please check other parts of the code without a brief description of some functions.